lib/rgeo/cartesian/feature_methods.rb

```diff
@@ -50,31 +50,161 @@ def segments
+        # check coordinates are all valid
+        exterior_ring.points.each do |pt|
+          check = check_invalid_coordinate(pt)
+          return check unless check.nil?
+        end
+        interior_rings.each do |ring|
+          ring.points.each do |pt|
+            check = check_invalid_coordinate(pt)
+            return check unless check.nil?
+          end
+        end
+
+        # check closed
+        return ImplHelper::TopologyErrors::UNCLOSED_RING unless exterior_ring.is_closed?
+        return ImplHelper::TopologyErrors::UNCLOSED_RING unless interior_rings.all?(&:is_closed?)
+
+        # check more than 3 points in each ring
+        return ImplHelper::TopologyErrors::TOO_FEW_POINTS unless exterior_ring.num_points > 3
+        return ImplHelper::TopologyErrors::TOO_FEW_POINTS unless interior_rings.all? { |r| r.num_points > 3 }
+
+        # can skip this check if there's no holes
+        if interior_rings.size.positive?
+          check = check_consistent_area
+          return check unless check.nil?
+        end
+
+        # check that there are no self-intersections
+        check = check_no_self_intersecting_rings(self)
+        return check unless check.nil?
+
+        # can skip these checks if there's no holes
+        if interior_rings.size.positive?
+          check = check_holes_in_shell(self)
+          return check unless check.nil?
+
+          check = check_holes_not_nested(self)
+          return check unless check.nil?
+
+          check = check_connected_interiors
+          return check unless check.nil?
+        end
+
+        nil
+      end
+
+      # Checks that there are no invalid intersections between the components
+      # of a polygon.
+      #
+      # @return [String] invalid_reason
+      def check_consistent_area
+        # Get set of unique coords
+        pts = exterior_ring.coordinates.to_set
+        interior_rings.each do |ring|
+          pts += ring.coordinates
+        end
+        num_points = pts.size
+
+        # if additional nodes were added, there must be an intersection
+        # through a boundary.
+        if graph.incident_edges.size > num_points
+          return ImplHelper::TopologyErrors::SELF_INTERSECTION
+        end
+
+        rings = [exterior_ring] + interior_rings
+        return ImplHelper::TopologyErrors::DUPLICATE_RINGS if rings.uniq.size != rings.size
+
+        nil
+      end
+
+      # Checks that the interior of a polygon is connected.
+      #
+      # Process to do this is to walk around an interior cycle of the
+      # exterior shell in the polygon's geometry graph. It will keep track
+      # of all the nodes it visited and if that set is a superset of the
+      # coordinates in the exterior_ring, the interior is connected, otherwise
+      # it is split somewhere.
+      #
+      # @return [String] invalid_reason
+      def check_connected_interiors
+        exterior_coords = exterior_ring.coordinates.to_set
+
+        visited = []
```

Comment:
```suggestion
        visited = Set.new
```

just avoids an array object for nothing :)